### PR TITLE
UCP: Migrate scalar function `substring_index` from TiDB

### LINKED ISSUE #6955: UCP: Migrate scalar function `SubstringIndex` from TiDB
## Description

Port the scalar function `SubstringIndex` from TiDB to coprocessor.

## Score 

* 50

## Mentor(s)

* @breeswish 

## Recommended Skills

* Rust programming 

## Learning Materials

Already implemented expressions ported from TiDB

- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/rpn_expr) 
- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/expr)

### PR DESCRIPTION
**Pull request by longfangsong**
Signed-off-by: longfangsong <longfangsong@icloud.com>

UCP: #6955

### What problem does this PR solve?

Issue Number: close #6955

Problem Summary: Migrate scalar function `substring_index` from TiDB

### What is changed and how it works?

What's Changed:

Implemented scalar function `substring_index`

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### CODE REVIEW

**Comment by breeswish** [comment]
@iosmanthus PTAL

File: components/tidb_query_vec_expr/src/impl_string.rs [hunk 1]
```diff
@@ -452,6 +453,71 @@ pub fn space(len: &Option<Int>) -> Result<Option<Bytes>> {
...
 
+#[rpn_fn]
+#[inline]
+pub fn substring_index(
+    s: &Option<Bytes>,
+    delim: &Option<Bytes>,
+    count: &Option<Int>,
+) -> Result<Option<Bytes>> {
+    Ok(
+        if let (Some(s), Some(delim), Some(count)) = (s, delim, count) {
```
Comment (iosmanthus) [inline components/tidb_query_vec_expr/src/impl_string.rs:464]: Seems we can unify the logic in `Greeter` arm and `Less` arm.
```rust
    if count == 0 || s.len() == 0 || delim.len() == 0 {
        return Vec::new();
    }

    let finder = if count > 0 {
        twoway::find_bytes
    } else {
        twoway::rfind_bytes
    };

    let mut r = &s[..];
    let mut i = count.abs();
    let mut bound = 0;

    while i > 0 {
        if let Some(index) = finder(r, &delim) {
            if count > 0 {
                bound += index + 1;
                r = &s[bound..];
            } else {
                bound = index;
                r = &s[..bound];
            }
        } else {
            break;
        }
        i -= 1;
    }

    if i > 0 {
        s.to_vec()
    } else {
        if count > 0 {
            s[..bound - 1].to_vec()
        } else {
            s[bound + 1..].to_vec()
        }
    }
```

**Comment by iosmanthus** [comment]
@longfangsong please handle the conflicts 😊

**Comment by iosmanthus** [comment]
@longfangsong Please handle the conflicts. @zhongzc PTAL.

**Pull request merged**